### ISSUE TITLE
tokenize_and_cut function

### ISSUE THREAD

**Issue by antgr** [bentrevett/pytorch-sentiment-analysis#55]
Hi,
thank you for this tutorial. Your work is amazing. Simple, elegant, crystal clear and truly useful,I hope that I new about your tutorials earlier, but it is never too late.
Regarding the current issue:
In tokenize_and_cut(..) there is only one argument:
```def tokenize_and_cut(sentence):```
but it uses tokenizer as well. This works because tokenizer is in the global namespace. On the other hand if tokenizer was an extra argument, then I do not know if that would be problem for passing this method as the custom tokenization in 
```data.Field( ... , tokenize = tokenize_and_cut, .. ```
Seems that would be.
Any comment on that?

**Comment by bentrevett** [comment]
I believe the way to solve this is to use `partial`

```python
import functools

def tokenize_and_cut(tokenizer, max_length, sentence):
    tokens = tokenizer.tokenize(sentence) 
    tokens = tokens[:max_input_length-2]
    return tokens

field_tokenizer = functools.partial(tokenize_and_cut(tokenizer, max_length))

TEXT = data.Field(tokenize = field_tokenizer, ...
```

Not at my computer right now so can't double check this works, but I'm pretty sure it's OK.

The reason I abused the global namespace is because I wanted to focus more on the explanation of how things are implemented. I may change my mind on this soon though. TorchText is going through some big changes in regards to how it handles datasets so I might re-write large portions of these tutorials.

**Comment by antgr** [comment]
ok thanks!

**Comment by bentrevett** [comment]
I don't believe there is a standard way as of yet - but it's something that can be hacked together. 

Are you wanting to train on the two datasets simultaneously or alternate between updates?

**Comment by bentrevett** [comment]
OK I saw your issue posted. If you want to loop through every example in each of the iterators but choose randomly from each until it's done and then only pick from the other then something like this should work:

```python
import random

x = [1,2,3,4,5]
y = ['a','b','c','d','e']

iter_x = iter(x)
iter_y = iter(y)

x_done = False
y_done = False

while True:
        
    if x_done and y_done:
        break
    elif y_done and not x_done:
        choice = 1
    elif x_done and not y_done:
        choice = 0
    else:
        choice = random.random()
    
    if choice > 0.5:
        #get from x
        try:
            batch = next(iter_x)
            #train
        except StopIteration:
            x_done = True
        
    else:
        #get from y
        try:
            batch = next(iter_y)
            #train
        except StopIteration:
            y_done = True
```

Obviously you'd use the iterators and not lists, but this is just an example.

**Comment by antgr** [comment]
> I don't believe there is a standard way as of yet - but it's something that can be hacked together.
> 
> Are you wanting to train on the two datasets simultaneously or alternate between updates?

I do not know what is needed for succesfull training. The description I gave was based on my current understanding, as a starting approach, and thanks for the example! But based on your question (simultaneously or alternate training between updates), seems that there maybe exist a better way. Probably between updates could give more stable results. I am not sure I understand exactly what means between updates, I will study again the training loop.

**Comment by bentrevett** [comment]
By training simultaneously I mean something like your model is trying (for example) to both predict sentiment and the POS tags for each word in a movie review. Thus your model accepts one input (the text) but has multiple outputs (both the tags and sentiments). Hence you have one model that is trying to do two tasks.

By alternating between updates I mean something like you are training a model to predict sentiment of a movie review and predict which genre a book is from it's description. Hence you have one model trained on two distinct datasets with two different tasks.

From the sound if it, I think you're wanting to do the second one?

**Comment by antgr** [comment]
Yes, the second one! Because what I am thinking and from what I have read, it uses information from two sources, which could be beneficial. (This doesn't mean that the first couldn't be helpful mainly if the first task is a subtask of the second, because in that case I suspect that you could help the training, but the problem here is that you probably need labels for both tasks, and I do not think that there are many free datasets for two or more tasks. To overcome this, the other solution would be to have an unsupervised task). So yes I want the second.

**Comment by bentrevett** [comment]
Yeah, I don't think I know of any datasets with labels for multiple task. I have seen them use an unsupervised task, such as language modelling though - I was considering adding a tutorial on it to introduce multi-task learning. I haven't properly researched into it though, so no idea how much of a benefit we'd gain from it.

**Comment by antgr** [comment]
From your tutorial we will benefit without doubt. From multitask learning, we could benefit only if the tasks would be related. For example knowing the POS will not help to understand the sentiment. But a model for example on NLI could help with the sentiment more.  Especially if you could analyze the text on its sentence pairs and take a score. But practically this could be complicated, I do not have the experience to evaluate the difficulty. These are just some thoughts.

**Comment by antgr** [comment]
I could close this, unless you want it open.

**Comment by bentrevett** [comment]
I’ll close it.

**Comment by antgr** [comment]
Regarding the talk I opened for multitask learning, I made an experiment, training sentiment analysis, together with nli. I could share the code and results.

**Comment by bentrevett** [comment]
@antgr I would definitely be interested in seeing your results!

**Comment by antgr** [comment]
I only ran it once. These are the results:
```
-------------------------------------
Epoch: 10 | Epoch Time: 7m 43s
Train Loss: 0.171 | Train Acc: 93.29%
Val. Loss: 0.228 |  Val. Acc: 91.82% <=== compare with a
Test Loss: 0.201 | Test Acc: 92.30% <==== compare with b
we now have:
----------------------------------------------
Epoch: 10 | Epoch Time: 8m 9s
Train Loss: 0.987 | Train Acc: 157.63%   --- OK here I have a bug :)
Val. Loss sent: 0.210 |  Val. Acc: 92.08%   <=== (a) small benefit for sentiment analysis when adding nli
Val. Loss nli: 0.824 |  Val. Acc: 62.79%  <=== nli sees degradation in its performance 
Test Loss sent: 0.190 | Test Acc: 92.46%  <==== (b) small benefit for sentiment analysis when adding nli
Test Loss sent : 0.826 | Test Acc: 63.07%
```

**Comment by bentrevett** [comment]
Looks good! Is this trained with the IMDB dataset? How are you getting NLI and sentiment analysis labels on the same dataset?

**Comment by antgr** [comment]
The network sees one example of IMDB and one example of SNLI, etc until one of the datasets has no other examples.

**Comment by antgr** [comment]
@bentrevett Do you think that we could try something better? Thinking out loud, the sentence of NLI *could* have a degree of sentiment (but seeing some of them, I think that they are not loaded with sentiment). Also, to the opposite direction, if we could take pair of sentences of imdb, we could label them as 'entails' or 'contradicts 'if they have the same or different labels (here it would be better if we had three classes as sentiments. in order to produce also the neutral label). But this would be a different meaning of 'entails' and 'contradicts' and 'neutral' compared to nli's.  I would try this, but I to be honest I do not think that it will help. I could do that just for curiosity.

**Comment by bentrevett** [comment]
> the sentence of NLI _could_ have a degree of sentiment
Potentially, yes. See: https://openai.com/blog/unsupervised-sentiment-neuron/, where they trained a language model that implicitly learns sentiment as it helps predict the following tokens in text. However, their model was trained on Amazon reviews which do explicitly have some degree of sentiment (positive for a good review, negative for a bad review) whereas I doubt the SNLI dataset has anything like that.

Following on from that, if we do something like you mentioned pair random IMDB reviews and say they 'entail' if their labelled sentiment is the same and 'contradict' if their sentiment is different, we might get a result like the OpenAI work. But, like you said, I am unsure if this will even give us any benefit in terms of prediction accuracy though as I think ideally we want the NLI dataset to be different from the sentiment analysis dataset. The OpenAI work mentions how it has learned this sentiment through mass amounts of data, and if we are just implicitly learning it from the dataset we are going to use for sentiment analysis anyway we hope our model would just learn it whilst learning to predict sentiment.

One idea is to use the other sentiment analysis datasets in TorchText, e.g. the AmazonReview or YelpReview, to create our fake NLI dataset. These are distinct from the IMDB dataset but do contain sentiment within them. I don't think a 'neutral' class is necessary in this case.

That's just a spitball of ideas, let me know what you think.

**Comment by antgr** [comment]
Thanks for sharing your thoughts! 
Definitely using other sentiment analysis datasets is a straight forward solution. I do not see how this would not help! 
What I think also had in mind when I proposed to infer a new weak inference task from sentiment analysis task, was to help nli task as well (the other direction). But now I see it again, and after your answer, does not seem good idea, because all these cases from nli's point of view, should be neutral anyway. 

One extra idea to the table. If we could add one more task, the task of 'sentence similarity' then maybe we would hope that the approximated equation  : *contradiction = opposite sentiment + high sentence similarity * could be proven helpful.

**Comment by antgr** [comment]
I hope that I didn't cause too much chaos in the discussion!

**Comment by bentrevett** [comment]
Going the other way may (training on movie reviews to improve NLI performance) may be easier. Again, the problem with the data not being in the same domain is what is going to cause most of the issues. 

I am assuming you are using the SNLI dataset. The examples in that dataset are built from image captions, i.e. an example of an entailment within the dataset is between "A soccer game with multiple males playing." and "Some men are playing a sport.". What will the neural network learn from this? Hopefully "multiple males" and "men" are the same thing and that "soccer" is a "sport". Now let's look at two positive examples in the Amazon reviews dataset: 

- "The creepiest book I've ever read! It's a creepy mystery/romance all in one creepy book! I would read it again lol"
- "It starts off a bit slow, but once the product placement jokes start it takes off."

What would a neural network learn in an NLI task on these examples? Nothing on here is really like the "soccer = sport" example above, it has to learn that "I would read it again" indicates a positive review and "it takes off" indicates a positive review. So could we say that it sort of has to learn sentiment (via learning if the review contains any phrases indicative of sentiment) to do the NLI task here? And if it does do that, surely this will help the neural net to predict sentiment? 

If learning NLI on reviews implicitly has to learn sentiment, then surely we can do things the other way around? If we learn on reviews to discover positive sentiment, then surely using that to just compare the predicted sentiment between reviews is a good way to do an NLI task on reviews? Do we need to even train on any of the examples in an NLI way? I guess if you present the network with reviews A and B, where A is positive and B is negative, and it incorrectly assumes B is positive and that the two examples are an entailment, then the NLI task may give us some sort of signal for learning that we are incorrect and our representations of at least one of the sentences is not correct.

There's a lot to think here, it definitely deserves more research and a lot of experiments to be run.
